File: app/src/main/java/com/chopcode/rutago/app/viewmodels/passenger/UserProfileViewModel.java
package com.chopcode.rutago.app.viewmodels.passenger;

import android.net.Uri;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.chopcode.rutago.app.config.MyApp;
import com.chopcode.rutago.app.managers.core.auth.AuthManager;
import com.chopcode.rutago.app.models.User;
import com.chopcode.rutago.app.services.reservations.passenger.PassengerReservationService;
import com.chopcode.rutago.app.services.storage.StorageService;
import com.chopcode.rutago.app.services.user.UserService;
import com.google.firebase.database.ValueEventListener;

import java.util.Map;

/**
 * User Profile ViewModel (Passenger)
 *
 * Gestor reactivo para la identidad y métricas personales del pasajero.
 * Responsabilidades:
 * - Mantener una suscripción en tiempo real al perfil del usuario en /usuarios/.
 * - Orquestar el cálculo de estadísticas de fidelización (Premium Stats).
 * - Gestionar la carga de avatares a Firebase Storage y su posterior vinculación a la DB.
 * - Implementar el flujo legal de eliminación de cuenta y cambio de estado de actividad.
 * - Sincronizar el ID de usuario autenticado mediante el AuthManager.
 */
public class UserProfileViewModel extends ViewModel {
    private static final String TAG = "UserProfileVM";

    private final MutableLiveData<User> userData = new MutableLiveData<>();
    
    /** Métricas acumuladas de fidelización (Inversión total, Puntos, Rutas favoritas). */
    private final MutableLiveData<Map<String, Object>> premiumStats = new MutableLiveData<>();
    
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<Boolean> isStatsLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> uploadStatus = new MutableLiveData<>();
    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final MutableLiveData<Boolean> accountDeletionSuccess = new MutableLiveData<>(false);

    private final UserService userService;
    private final StorageService storageService;
    private final PassengerReservationService passengerReservationService;
    private final AuthManager authManager;
    private ValueEventListener userListener;

    public UserProfileViewModel() {
        this.userService = new UserService();
        this.storageService = new StorageService();
        this.passengerReservationService = new PassengerReservationService();
        this.authManager = AuthManager.getInstance();
    }

    public LiveData<User> getUserData() { return userData; }
    public LiveData<Map<String, Object>> getPremiumStats() { return premiumStats; }
    public LiveData<Boolean> getIsLoading() { return isLoading; }
    public LiveData<Boolean> getIsStatsLoading() { return isStatsLoading; }
    public LiveData<String> getUploadStatus() { return uploadStatus; }
    public LiveData<String> getError() { return error; }
    public LiveData<Boolean> getAccountDeletionSuccess() { return accountDeletionSuccess; }

    /**
     * Activa la escucha reactiva del perfil del pasajero.
     * Al cargar los datos, dispara automáticamente la actualización de estadísticas.
     */
    public void loadProfile() {
        String userId = authManager.getUserId();
        if (userId == null) return;

        if (userListener != null) return;

        isLoading.setValue(true);
        userListener = userService.listenToUserData(userId, new UserService.UserDataCallback() {
            @Override public void onUserDataLoaded(User user) { 
                user.setId(userId);
                userData.postValue(user); 
                isLoading.postValue(false); 
                loadPremiumStats(userId); 
            }
            @Override public void onError(String errorMsg) { error.postValue(errorMsg); isLoading.postValue(false); }
        });
    }

    /**
     * Calcula métricas de uso históricas para el programa de fidelización.
     */
    private void loadPremiumStats(String userId) {
        isStatsLoading.setValue(true);
        passengerReservationService.getPremiumStats(userId, new PassengerReservationService.PremiumStatsCallback() {
            @Override public void onStatsCalculated(Map<String, Object> stats) { premiumStats.postValue(stats); isStatsLoading.postValue(false); }
            @Override public void onError(String errorMsg) { Log.e(TAG, "❌ Error al calcular estadísticas: " + errorMsg); isStatsLoading.postValue(false); }
        });
    }

    /**
     * Sube una nueva fotografía de perfil.
     */
    public void uploadProfilePicture(Uri uri) {
        String userId = authManager.getUserId();
        if (userId == null) return;
        uploadStatus.setValue("Uploading...");
        storageService.uploadProfilePicture(userId, uri, new StorageService.UploadCallback() {
            @Override public void onSuccess(String downloadUrl) { updateProfilePictureUrl(userId, downloadUrl); }
            @Override public void onError(String errorMsg) { error.postValue("Upload error: " + errorMsg); uploadStatus.postValue(null); }
            @Override public void onProgress(double progress) {}
        });
    }

    private void updateProfilePictureUrl(String userId, String downloadUrl) {
        userService.updateProfilePicture(userId, downloadUrl, "usuarios", new UserService.UserUpdateCallback() {
            @Override public void onSuccess() { uploadStatus.postValue("Updated"); }
            @Override public void onError(String errorMsg) { error.postValue("DB error: " + errorMsg); uploadStatus.postValue(null); }
        });
    }

    /**
     * Inicia el proceso de borrado con periodo de gracia.
     */
    public void requestAccountDeletion() {
        String userId = authManager.getUserId();
        if (userId == null) return;
        userService.requestAccountDeletion(userId, "usuarios", new UserService.UserUpdateCallback() {
            @Override public void onSuccess() { accountDeletionSuccess.postValue(true); }
            @Override public void onError(String errorMsg) { error.postValue(errorMsg); }
        });
    }

    /**
     * Reinstaura una cuenta que estaba marcada para eliminación.
     */
    public void cancelAccountDeletion() {
        String userId = authManager.getUserId();
        if (userId == null) return;
        isLoading.setValue(true);
        userService.cancelAccountDeletion(userId, "usuarios", new UserService.UserUpdateCallback() {
            @Override public void onSuccess() { isLoading.postValue(false); }
            @Override public void onError(String errorMsg) { error.postValue(errorMsg); isLoading.postValue(false); }
        });
    }

    /**
     * Cambia el estado del pasajero (Activo/Inactivo).
     * No disponible si la cuenta está bloqueada administrativamente.
     */
    public void toggleUserStatus() {
        User current = userData.getValue();
        if (current == null || "blocked".equals(current.getStatus())) return;
        String newStatus = "active".equals(current.getStatus()) ? "inactive" : "active";
        String userId = authManager.getUserId();
        if (userId == null) return;
        isLoading.postValue(true);
        userService.updateUserStatus(userId, newStatus, new UserService.UserUpdateCallback() {
            @Override public void onSuccess() { isLoading.postValue(false); }
            @Override public void onError(String errorMsg) { error.postValue(errorMsg); isLoading.postValue(false); }
        });
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (userListener != null) {
            String uid = authManager.getUserId();
            if (uid != null) {
                MyApp.getDatabaseReference("usuarios/" + uid).removeEventListener(userListener);
                Log.d(TAG, "🧹 Listener de perfil de pasajero removido.");
            }
        }
    }
}

File: app/src/main/java/com/chopcode/rutago/app/ui/viewmodels/passenger/UserProfileViewModel.kt
package com.chopcode.rutago.app.ui.viewmodels.passenger

import android.net.Uri
import androidx.lifecycle.ViewModel
import com.chopcode.rutago.app.config.MyApp
import com.chopcode.rutago.app.managers.core.auth.AuthManager
import com.chopcode.rutago.app.models.User
import com.chopcode.rutago.app.services.reservations.passenger.PassengerReservationService
import com.chopcode.rutago.app.services.storage.StorageService
import com.chopcode.rutago.app.services.user.UserService
import com.chopcode.rutago.app.data.repositories.settings.SettingsRepository
import com.chopcode.rutago.app.data.repositories.settings.SettingsRepositoryImpl
import com.google.firebase.database.ValueEventListener
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * 🧠 VIEWMODEL: UserProfileViewModel
 * Gestor reactivo para la identidad y métricas personales del pasajero.
 */
class UserProfileViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(UserProfileUiState())
    val uiState: StateFlow<UserProfileUiState> = _uiState.asStateFlow()

    private val userService = UserService()
    private val storageService = StorageService()
    private val reservationService = PassengerReservationService()
    private val authManager = AuthManager.getInstance()
    private val settingsRepository: SettingsRepository = SettingsRepositoryImpl(MyApp.getAppContext())
    private var userListener: ValueEventListener? = null

    init {
        loadProfile()
        checkTutorial()
    }

    private fun checkTutorial() {
        if (settingsRepository.shouldShowTutorial("tut_profile")) {
            _uiState.update { it.copy(showTutorial = true) }
        }
    }

    fun onTutorialDismiss() {
        settingsRepository.markTutorialAsSeen("tut_profile")
        _uiState.update { it.copy(showTutorial = false) }
    }

    fun loadProfile() {
        val userId = authManager.userId ?: return
        if (userListener != null) return

        _uiState.update { it.copy(isLoading = true) }

        userListener = userService.listenToUserData(userId, object : UserService.UserDataCallback {
            override fun onUserDataLoaded(user: User?) {
                user?.let { u ->
                    u.id = userId
                    _uiState.update { it.copy(user = u, isLoading = false) }
                    loadPremiumStats(userId)
                }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(isLoading = false, error = errorMsg) }
            }
        })
    }

    private fun loadPremiumStats(userId: String) {
        _uiState.update { it.copy(isStatsLoading = true) }
        reservationService.getPremiumStats(userId, object : PassengerReservationService.PremiumStatsCallback {
            override fun onStatsCalculated(stats: Map<String, Any>?) {
                stats?.let { s ->
                    _uiState.update { 
                        it.copy(
                            totalSpent = s["totalSpent"]?.toString() ?: "$0",
                            loyaltyPoints = "${s["loyaltyPoints"]} pts",
                            favoriteRoute = s["favoriteRoute"]?.toString() ?: "No disponible",
                            isStatsLoading = false
                        )
                    }
                }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(isStatsLoading = false) }
            }
        })
    }

    fun uploadProfilePicture(uri: Uri) {
        val userId = authManager.userId ?: return
        _uiState.update { it.copy(uploadStatus = "Uploading...") }
        
        storageService.uploadProfilePicture(userId, uri, object : StorageService.UploadCallback {
            override fun onSuccess(downloadUrl: String?) {
                updateProfilePictureUrl(userId, downloadUrl)
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(error = "Upload error: $errorMsg", uploadStatus = null) }
            }

            override fun onProgress(progress: Double) {}
        })
    }

    private fun updateProfilePictureUrl(userId: String, downloadUrl: String?) {
        userService.updateProfilePicture(userId, downloadUrl, "usuarios", object : UserService.UserUpdateCallback {
            override fun onSuccess() {
                _uiState.update { it.copy(uploadStatus = "Updated") }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(error = "DB error: $errorMsg", uploadStatus = null) }
            }
        })
    }

    fun toggleUserStatus() {
        val currentUser = _uiState.value.user ?: return
        if (currentUser.status == "blocked") return
        
        val newStatus = if (currentUser.status == "active") "inactive" else "active"
        val userId = authManager.userId ?: return
        
        _uiState.update { it.copy(isLoading = true) }
        userService.updateUserStatus(userId, newStatus, object : UserService.UserUpdateCallback {
            override fun onSuccess() {
                _uiState.update { it.copy(isLoading = false) }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(error = errorMsg, isLoading = false) }
            }
        })
    }

    fun requestAccountDeletion() {
        val userId = authManager.userId ?: return
        userService.requestAccountDeletion(userId, "usuarios", object : UserService.UserUpdateCallback {
            override fun onSuccess() {
                _uiState.update { it.copy(accountDeletionSuccess = true) }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(error = errorMsg) }
            }
        })
    }

    fun cancelAccountDeletion() {
        val userId = authManager.userId ?: return
        _uiState.update { it.copy(isLoading = true) }
        userService.cancelAccountDeletion(userId, "usuarios", object : UserService.UserUpdateCallback {
            override fun onSuccess() {
                _uiState.update { it.copy(isLoading = false) }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(error = errorMsg, isLoading = false) }
            }
        })
    }

    fun showDeleteConfirmation(show: Boolean) {
        _uiState.update { it.copy(showDeleteConfirmation = show) }
    }

    fun onLogoutClick() {
        _uiState.update { it.copy(showLogoutDialog = true) }
    }

    fun dismissLogoutDialog() {
        _uiState.update { it.copy(showLogoutDialog = false) }
    }

    fun logout() {
        com.google.firebase.auth.FirebaseAuth.getInstance().signOut()
        _uiState.update { it.copy(showLogoutDialog = false, logoutSuccess = true) }
    }

    override fun onCleared() {
        super.onCleared()
        userListener?.let {
            authManager.userId?.let { uid ->
                MyApp.getDatabaseReference("usuarios/$uid").removeEventListener(it)
            }
        }
    }
}
